### ISSUE TITLE
Consider using std::tanh for sigmoid

### ISSUE THREAD

**Issue by Kamairo** [tiny-dnn/tiny-dnn#412]
After a few runtime tests, I came to the conlusion, that `std::tanh` evaluates much faster than `std::exp`. Since sigmoid is a scaled and moved version of tanh, I would suggest you should use
```cpp
    float_t f(const vec_t& v, cnn_size_t i) const override { return float_t(0.5) * ( float_t(1.0) + std::tanh(v[i] * float_t(0.5))); }
```
instead of
```cpp
    float_t f(const vec_t& v, cnn_size_t i) const override { return float_t(1) / (float_t(1) + std::exp(-v[i])); }
```
which evaluates much, much faster in my tests.

**Comment by bhack** [comment]
You can also add a fast sigmoid http://image.diku.dk/shark/sphinx_pages/build/html/rest_sources/tutorials/algorithms/ffnet.html

**Comment by bhack** [comment]
And also https://www.quora.com/What-is-hard-sigmoid-in-artificial-neural-networks-Why-is-it-faster-than-standard-sigmoid-Are-there-any-disadvantages-over-the-standard-sigmoid

**Comment by Kamairo** [comment]
Good point, but both of these solutions are approximations, while using the tanh function as described above is not (as far as numerical limits go).

**Comment by bhack** [comment]
It is hard to tell if it is faster on each archs see first and second replies http://stackoverflow.com/questions/10732027/fast-sigmoid-algorithm

**Comment by nyanp** [comment]
@Kamairo 
Nice idea :)

We can see glibc implementations:
- [expf](https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/ieee754/flt-32/e_expf.c;h=071f615ef40c9222b7d56d247b12bc1530dfe86a;hb=HEAD)
- [tanhf](https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/ieee754/flt-32/s_tanhf.c;h=f70702b29c3c43446d8ac89e1d6628e827cacd3f;hb=HEAD)

It is interesting that ```tanhf``` use very fast approximation when x is close to zero. If the outputs of the network is sparse, this could be advantageous to use ```tanh``` instead of ```exp```.

```cpp
if (ix == 0)
    return x;               /* x == +-0 */
if (ix<0x24000000)          /* |x|<2**-55 */
{
    math_check_force_underflow (x);
    return x*(one+x);       /* tanh(small) = small */
}
```

In normal case, the performance of ```tanh``` may be similar to ```expf```, because ```tanh``` calls ```expm1f```. As you can see [this comment](https://gist.github.com/astanin/5270668#gistcomment-1491696), its performance heavily depends on its input value.

The best way is to measure the performance of the training real networks. Does anyone have similar experience?
